sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotInstantiateSharedClasses.cs

```diff
@@ -0,0 +1,85 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotInstantiateSharedClasses : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4277";
+        private const string MessageFormat = "Refactor this code so that it doesn't invoke the constructor of this class.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    if (c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var creationSyntax = (ObjectCreationExpressionSyntax)c.Node;
+
+                    var createdType = c.SemanticModel.GetTypeInfo(creationSyntax).Type;
+                    if (createdType != null &&
+                        createdType.GetAttributes().Any(IsSharedPartCreationPolicyAttribute))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, creationSyntax.GetLocation()));
+                    }
+                },
+                SyntaxKind.ObjectCreationExpression);
+        }
+
+        private static bool IsSharedPartCreationPolicyAttribute(AttributeData data)
+        {
+            return IsPartCreationPolicyAttribute(data) && IsShared(data);
+        }
+
+        private static bool IsPartCreationPolicyAttribute(AttributeData data)
+        {
+            return data.AttributeClass.Is(KnownType.System_ComponentModel_Composition_PartCreationPolicyAttribute);
+        }
+
+        private static bool IsShared(AttributeData data)
+        {
+            // This is equivalent to System.ComponentModel.Composition.CreationPolicy.Shared,
+            // but we do not want dependency on System.ComponentModel.Composition just for that.
+            const int CreationPolicy_Shared = 1;
```

Comment:
This can do but I think @duncanp-sonar was more thinking about comparing the full string rather than the value (i.e. `System.ComponentModel.CreationPolicy.Shared`)

Reply:
I was thinking of the full string (assuming it's available at runtime), but in practice it probably doesn't make much difference.

Reply:
And by the way I don't know which way is the best.

Reply:
I'd prefer the full string too, but it doesn't seem to be available. At least we couldn't find it with Amaury.